File: demoCode/raceConditions/sumTo10.c
#include <stdio.h>
#include <omp.h>

int main(){

  int sum = 0;
  for (int i=1; i<=10; i++){
    sum += i;
  }

  printf("without multithreading, the sum of first 10 integers is %d\n", sum);

  sum = 0;

  #pragma omp parallel for
  for (int i=1; i<=10; i++){
    sum += i;
  }

  printf("using multithreading, the sum of first 10 integers is %d\n", sum);

  sum = 0;

  #pragma omp parallel for
  for (int i=1; i<=10; i++){
    #pragma omp critical //one at a time through here
    sum += i;
  }

  printf("using multithreading with \"critical\" guard, the sum of first 10 integers is %d\n", sum);

  sum = 0;
  
  #pragma omp parallel for reduction(+:sum)
  for (int i=1; i<=10; i++){
    sum += i;
  }

  printf("using multithreading with reduction, the sum of first 10 integers is %d\n", sum);

}

File: omp/examples/raceConditions/sumTo10.f90
PROGRAM sumTo10
   USE omp_lib
   IMPLICIT NONE

   INTEGER :: sum, i

   sum = 0
   DO i = 1, 10
      sum = sum + i
   END DO
   WRITE(*,'(A,I0)') 'without multithreading, the sum of first 10 integers is                      ', sum

   sum = 0
   !$OMP PARALLEL DO
   DO i = 1, 10
      sum = sum + i
   END DO
   !$OMP END PARALLEL DO
   WRITE(*,'(A,I0)') 'using unsafe multithreading, the sum of first 10 integers is                 ', sum

   sum = 0
   !$OMP PARALLEL DO
   DO i = 1, 10
      !$OMP CRITICAL
      sum = sum + i   ! one at a time through here
      !$OMP END CRITICAL
   END DO
   !$OMP END PARALLEL DO
   WRITE(*,'(A,I0)') 'using multithreading with "critical" guard, the sum of first 10 integers is  ', sum

   sum = 0
   !$OMP PARALLEL DO REDUCTION(+:sum)
   DO i = 1, 10
      sum = sum + i
   END DO
   !$OMP END PARALLEL DO
   WRITE(*,'(A,I0)') 'using multithreading with reduction, the sum of first 10 integers is         ', sum

END PROGRAM sumTo10